File: examples/dom_example.cpp
// LightUI DOM API C++ 示例
// 演示如何使用 DOM API 构建和操作 DOM 树

#include <iostream>
#include <memory>
#include "core/dom/document.h"
#include "core/dom/element.h"
#include "core/dom/text.h"
#include "core/dom/event.h"

using namespace lightui;

// ========== 示例 1: 创建简单的 DOM 结构 ==========

void Example1_CreateSimpleDOM() {
    std::cout << "=== 示例 1: 创建简单的 DOM 结构 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    
    // 创建容器
    auto container = doc->CreateElement("div");
    container->SetAttribute("id", "container");
    container->SetAttribute("class", "main-container");
    
    // 创建标题
    auto title = doc->CreateElement("h1");
    title->SetAttribute("id", "title");
    title->AppendChild(doc->CreateTextNode("Welcome to LightUI"));
    
    // 创建段落
    auto paragraph = doc->CreateElement("p");
    paragraph->SetAttribute("class", "intro");
    paragraph->AppendChild(doc->CreateTextNode("This is a lightweight UI framework."));
    
    // 组装 DOM 树
    container->AppendChild(title);
    container->AppendChild(paragraph);
    
    std::cout << "Container ID: " << container->GetAttribute("id") << std::endl;
    std::cout << "Container class: " << container->GetAttribute("class") << std::endl;
    std::cout << std::endl;
}

// ========== 示例 2: 属性操作 ==========

void Example2_AttributeOperations() {
    std::cout << "=== 示例 2: 属性操作 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    auto button = doc->CreateElement("button");
    
    // 设置属性
    button->SetAttribute("id", "myButton");
    button->SetAttribute("class", "btn btn-primary");
    button->SetAttribute("data-action", "submit");
    button->SetAttribute("data-target", "#form");
    
    // 读取属性
    std::cout << "Button ID: " << button->GetAttribute("id") << std::endl;
    std::cout << "Button class: " << button->GetAttribute("class") << std::endl;
    std::cout << "Data action: " << button->GetAttribute("data-action") << std::endl;
    std::cout << "Data target: " << button->GetAttribute("data-target") << std::endl;
    
    // 检查属性
    if (button->HasAttribute("id")) {
        std::cout << "Button has ID attribute" << std::endl;
    }
    
    // 删除属性
    button->RemoveAttribute("data-target");
    std::cout << "After removing data-target: " << button->GetAttribute("data-target") << std::endl;
    std::cout << std::endl;
}

// ========== 示例 3: 构建列表 ==========

void Example3_BuildList() {
    std::cout << "=== 示例 3: 构建列表 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    auto list = doc->CreateElement("ul");
    list->SetAttribute("id", "itemList");
    list->SetAttribute("class", "list");
    
    // 创建 5 个列表项
    for (int i = 1; i <= 5; i++) {
        auto item = doc->CreateElement("li");
        item->SetAttribute("class", "list-item");
        item->SetAttribute("id", "item-" + std::to_string(i));
        item->AppendChild(doc->CreateTextNode("Item " + std::to_string(i)));
        list->AppendChild(item);
    }
    
    std::cout << "Created list with " << list->GetChildNodes().size() << " items" << std::endl;
    std::cout << std::endl;
}

// ========== 示例 4: 查询元素 ==========

void Example4_QueryElements() {
    std::cout << "=== 示例 4: 查询元素 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    
    // 创建一些元素
    auto container = doc->CreateElement("div");
    container->SetAttribute("id", "container");
    
    auto title = doc->CreateElement("h1");
    title->SetAttribute("id", "title");
    title->SetAttribute("class", "heading");
    container->AppendChild(title);
    
    auto para1 = doc->CreateElement("p");
    para1->SetAttribute("class", "text");
    container->AppendChild(para1);
    
    auto para2 = doc->CreateElement("p");
    para2->SetAttribute("class", "text");
    container->AppendChild(para2);
    
    // 通过 ID 查询（最快）
    auto foundTitle = doc->GetElementById("title");
    if (foundTitle) {
        std::cout << "Found title by ID: " << foundTitle->GetAttribute("id") << std::endl;
    }
    
    // 通过 QuerySelector 查询
    auto foundHeading = container->QuerySelector(".heading");
    if (foundHeading) {
        std::cout << "Found heading by class" << std::endl;
    }
    
    // 查询所有段落
    auto paragraphs = container->QuerySelectorAll(".text");
    std::cout << "Found " << paragraphs.size() << " paragraphs" << std::endl;
    std::cout << std::endl;
}

// ========== 示例 5: 事件处理 ==========

void Example5_EventHandling() {
    std::cout << "=== 示例 5: 事件处理 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    auto button = doc->CreateElement("button");
    button->SetAttribute("id", "myButton");
    
    // 添加点击事件监听器
    button->AddEventListener("click", [](std::shared_ptr<Event> e) {
        std::cout << "Button clicked!" << std::endl;
        std::cout << "Event type: " << e->GetType() << std::endl;
    });
    
    // 添加鼠标悬停事件
    button->AddEventListener("mouseover", [](std::shared_ptr<Event> e) {
        std::cout << "Mouse over button" << std::endl;
    });
    
    // 触发点击事件
    auto clickEvent = std::make_shared<Event>("click", true);
    button->DispatchEvent(clickEvent);
    
    // 触发鼠标悬停事件
    auto mouseEvent = std::make_shared<Event>("mouseover", true);
    button->DispatchEvent(mouseEvent);
    
    std::cout << std::endl;
}

// ========== 示例 6: 事件冒泡 ==========

void Example6_EventBubbling() {
    std::cout << "=== 示例 6: 事件冒泡 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    
    // 创建嵌套结构
    auto outer = doc->CreateElement("div");
    outer->SetAttribute("id", "outer");
    
    auto middle = doc->CreateElement("div");
    middle->SetAttribute("id", "middle");
    
    auto inner = doc->CreateElement("div");
    inner->SetAttribute("id", "inner");
    
    middle->AppendChild(inner);
    outer->AppendChild(middle);
    
    // 在每一层添加事件监听器
    outer->AddEventListener("click", [](std::shared_ptr<Event> e) {
        auto target = std::dynamic_pointer_cast<Element>(e->GetCurrentTarget());
        std::cout << "Outer div clicked, current target: " << target->GetAttribute("id") << std::endl;
    });
    
    middle->AddEventListener("click", [](std::shared_ptr<Event> e) {
        auto target = std::dynamic_pointer_cast<Element>(e->GetCurrentTarget());
        std::cout << "Middle div clicked, current target: " << target->GetAttribute("id") << std::endl;
    });
    
    inner->AddEventListener("click", [](std::shared_ptr<Event> e) {
        auto target = std::dynamic_pointer_cast<Element>(e->GetCurrentTarget());
        std::cout << "Inner div clicked, current target: " << target->GetAttribute("id") << std::endl;
    });
    
    // 在最内层触发事件
    auto event = std::make_shared<Event>("click", true);  // bubbles = true
    inner->DispatchEvent(event);
    
    std::cout << std::endl;
}

// ========== 示例 7: 卡片网格 ==========

void Example7_CardGrid() {
    std::cout << "=== 示例 7: 卡片网格 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    auto grid = doc->CreateElement("div");
    grid->SetAttribute("id", "cardGrid");
    grid->SetAttribute("class", "grid");
    
    // 创建 3x3 卡片网格
    for (int row = 0; row < 3; row++) {
        for (int col = 0; col < 3; col++) {
            auto card = doc->CreateElement("div");
            card->SetAttribute("class", "card");
            card->SetAttribute("id", "card-" + std::to_string(row) + "-" + std::to_string(col));
            
            auto cardTitle = doc->CreateElement("h3");
            cardTitle->AppendChild(doc->CreateTextNode("Card " + std::to_string(row * 3 + col + 1)));
            
            auto cardBody = doc->CreateElement("p");
            cardBody->AppendChild(doc->CreateTextNode("This is card content"));
            
            card->AppendChild(cardTitle);
            card->AppendChild(cardBody);
            grid->AppendChild(card);
        }
    }
    
    std::cout << "Created 3x3 card grid with " << grid->GetChildNodes().size() << " cards" << std::endl;
    
    // 查询所有卡片
    auto allCards = grid->QuerySelectorAll(".card");
    std::cout << "Found " << allCards.size() << " cards using QuerySelectorAll" << std::endl;
    
    // 查询特定卡片
    auto specificCard = grid->QuerySelector("#card-1-1");
    if (specificCard) {
        std::cout << "Found specific card: " << specificCard->GetAttribute("id") << std::endl;
    }
    
    std::cout << std::endl;
}

// ========== 示例 8: innerHTML ==========

void Example8_InnerHTML() {
    std::cout << "=== 示例 8: innerHTML ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    auto div = doc->CreateElement("div");
    
    auto h1 = doc->CreateElement("h1");
    h1->AppendChild(doc->CreateTextNode("Title"));
    
    auto p = doc->CreateElement("p");
    p->SetAttribute("class", "intro");
    p->AppendChild(doc->CreateTextNode("Introduction"));
    
    div->AppendChild(h1);
    div->AppendChild(p);
    
    // 获取 HTML 字符串
    std::string html = div->GetInnerHTML();
    std::cout << "innerHTML: " << html << std::endl;
    std::cout << std::endl;
}

// ========== 示例 9: 克隆节点 ==========

void Example9_CloneNode() {
    std::cout << "=== 示例 9: 克隆节点 ===" << std::endl;
    
    auto doc = std::make_shared<Document>();
    
    // 创建一个复杂的元素
    auto original = doc->CreateElement("div");
    original->SetAttribute("id", "original");
    original->SetAttribute("class", "container");
    
    auto child1 = doc->CreateElement("span");
    child1->AppendChild(doc->CreateTextNode("Child 1"));
    
    auto child2 = doc->CreateElement("span");
    child2->AppendChild(doc->CreateTextNode("Child 2"));
    
    original->AppendChild(child1);
    original->AppendChild(child2);
    
    // 浅克隆（不包括子节点）
    auto shallowClone = std::dynamic_pointer_cast<Element>(original->CloneNode(false));
    shallowClone->SetAttribute("id", "shallowClone");
    std::cout << "Shallow clone children: " << shallowClone->GetChildNodes().size() << std::endl;
    
    // 深克隆（包括所有子节点）
    auto deepClone = std::dynamic_pointer_cast<Element>(original->CloneNode(true));
    deepClone->SetAttribute("id", "deepClone");
    std::cout << "Deep clone children: " << deepClone->GetChildNodes().size() << std::endl;
    
    std::cout << std::endl;
}

// ========== 主函数 ==========

int main() {
    std::cout << "LightUI DOM API C++ 示例" << std::endl;
    std::cout << "=========================" << std::endl;
    std::cout << std::endl;
    
    Example1_CreateSimpleDOM();
    Example2_AttributeOperations();
    Example3_BuildList();
    Example4_QueryElements();
    Example5_EventHandling();
    Example6_EventBubbling();
    Example7_CardGrid();
    Example8_InnerHTML();
    Example9_CloneNode();
    
    std::cout << "=== 所有示例完成 ===" << std::endl;
    std::cout << "演示了以下功能:" << std::endl;
    std::cout << "1. 创建元素和文本节点" << std::endl;
    std::cout << "2. 设置和获取属性" << std::endl;
    std::cout << "3. 构建 DOM 树" << std::endl;
    std::cout << "4. 查询元素 (GetElementById, QuerySelector, QuerySelectorAll)" << std::endl;
    std::cout << "5. 事件处理和事件冒泡" << std::endl;
    std::cout << "6. 创建网格布局" << std::endl;
    std::cout << "7. innerHTML 操作" << std::endl;
    std::cout << "8. 克隆节点" << std::endl;
    
    return 0;
}

File: examples/dom_example.js
// LightUI DOM API JavaScript 示例
// 演示如何使用 DOM API 构建和操作 DOM 树

// ========== 示例 1: 创建简单的 DOM 结构 ==========

console.log("=== 示例 1: 创建简单的 DOM 结构 ===");

// 创建容器
var container = document.createElement('div');
container.id = 'container';
container.className = 'main-container';

// 创建标题
var title = document.createElement('h1');
title.id = 'title';
var titleText = document.createTextNode('Welcome to LightUI');
title.appendChild(titleText);

// 创建段落
var paragraph = document.createElement('p');
paragraph.className = 'intro';
var paraText = document.createTextNode('This is a lightweight UI framework.');
paragraph.appendChild(paraText);

// 组装 DOM 树
container.appendChild(title);
container.appendChild(paragraph);

console.log("Container created with ID:", container.id);
console.log("Container class:", container.className);

// ========== 示例 2: 属性操作 ==========

console.log("\n=== 示例 2: 属性操作 ===");

var button = document.createElement('button');
button.id = 'myButton';
button.className = 'btn btn-primary';

// 设置自定义属性
button.setAttribute('data-action', 'submit');
button.setAttribute('data-target', '#form');

// 读取属性
console.log("Button ID:", button.id);
console.log("Button class:", button.className);
console.log("Data action:", button.getAttribute('data-action'));
console.log("Data target:", button.getAttribute('data-target'));

// ========== 示例 3: 构建列表 ==========

console.log("\n=== 示例 3: 构建列表 ===");

var list = document.createElement('ul');
list.id = 'itemList';
list.className = 'list';

// 创建 5 个列表项
for (var i = 1; i <= 5; i++) {
    var item = document.createElement('li');
    item.className = 'list-item';
    item.id = 'item-' + i;
    
    var itemText = document.createTextNode('Item ' + i);
    item.appendChild(itemText);
    
    list.appendChild(item);
}

console.log("Created list with 5 items");

// ========== 示例 4: 查询元素 ==========

console.log("\n=== 示例 4: 查询元素 ===");

// 通过 ID 查询（最快）
var foundButton = document.getElementById('myButton');
if (foundButton) {
    console.log("Found button by ID:", foundButton.id);
}

// 通过 QuerySelector 查询
var foundTitle = container.querySelector('#title');
if (foundTitle) {
    console.log("Found title by QuerySelector");
}

// 查询所有列表项
var items = list.querySelectorAll('.list-item');
console.log("Found", items.length, "list items");

// ========== 示例 5: 事件处理 ==========

console.log("\n=== 示例 5: 事件处理 ===");

// 添加点击事件监听器
button.addEventListener('click', function(event) {
    console.log("Button clicked!");
    console.log("Event type:", event.type);
    console.log("Event target:", event.target.id);
});

// 添加鼠标悬停事件
button.addEventListener('mouseover', function(event) {
    console.log("Mouse over button");
});

console.log("Event listeners added to button");

// ========== 示例 6: 事件冒泡 ==========

console.log("\n=== 示例 6: 事件冒泡 ===");

var outer = document.createElement('div');
outer.id = 'outer';
outer.className = 'outer';

var middle = document.createElement('div');
middle.id = 'middle';
middle.className = 'middle';

var inner = document.createElement('div');
inner.id = 'inner';
inner.className = 'inner';

// 组装嵌套结构
middle.appendChild(inner);
outer.appendChild(middle);

// 在每一层添加事件监听器
outer.addEventListener('click', function(event) {
    console.log("Outer div clicked, current target:", event.currentTarget.id);
});

middle.addEventListener('click', function(event) {
    console.log("Middle div clicked, current target:", event.currentTarget.id);
});

inner.addEventListener('click', function(event) {
    console.log("Inner div clicked, current target:", event.currentTarget.id);
    // 可以停止冒泡
    // event.stopPropagation();
});

console.log("Event bubbling example set up");

// ========== 示例 7: 动态修改 DOM ==========

console.log("\n=== 示例 7: 动态修改 DOM ===");

var dynamicContainer = document.createElement('div');
dynamicContainer.id = 'dynamic';

// 添加初始内容
var initialText = document.createTextNode('Initial content');
dynamicContainer.appendChild(initialText);

console.log("Initial content added");

// 添加更多元素
var newPara = document.createElement('p');
var newText = document.createTextNode('Dynamically added paragraph');
newPara.appendChild(newText);
dynamicContainer.appendChild(newPara);

console.log("New paragraph added");

// 移除元素
// dynamicContainer.removeChild(newPara);
// console.log("Paragraph removed");

// ========== 示例 8: 表单元素 ==========

console.log("\n=== 示例 8: 表单元素 ===");

var form = document.createElement('form');
form.id = 'loginForm';
form.className = 'form';

// 用户名输入
var usernameLabel = document.createElement('label');
usernameLabel.setAttribute('for', 'username');
var labelText = document.createTextNode('Username:');
usernameLabel.appendChild(labelText);

var usernameInput = document.createElement('input');
usernameInput.id = 'username';
usernameInput.setAttribute('type', 'text');
usernameInput.setAttribute('name', 'username');
usernameInput.setAttribute('placeholder', 'Enter username');

// 密码输入
var passwordLabel = document.createElement('label');
passwordLabel.setAttribute('for', 'password');
var pwdLabelText = document.createTextNode('Password:');
passwordLabel.appendChild(pwdLabelText);

var passwordInput = document.createElement('input');
passwordInput.id = 'password';
passwordInput.setAttribute('type', 'password');
passwordInput.setAttribute('name', 'password');
passwordInput.setAttribute('placeholder', 'Enter password');

// 提交按钮
var submitButton = document.createElement('button');
submitButton.setAttribute('type', 'submit');
var submitText = document.createTextNode('Login');
submitButton.appendChild(submitText);

// 组装表单
form.appendChild(usernameLabel);
form.appendChild(usernameInput);
form.appendChild(passwordLabel);
form.appendChild(passwordInput);
form.appendChild(submitButton);

console.log("Login form created");

// ========== 示例 9: 卡片网格 ==========

console.log("\n=== 示例 9: 卡片网格 ===");

var grid = document.createElement('div');
grid.id = 'cardGrid';
grid.className = 'grid';

// 创建 3x3 卡片网格
for (var row = 0; row < 3; row++) {
    for (var col = 0; col < 3; col++) {
        var card = document.createElement('div');
        card.className = 'card';
        card.id = 'card-' + row + '-' + col;
        
        var cardTitle = document.createElement('h3');
        var cardTitleText = document.createTextNode('Card ' + (row * 3 + col + 1));
        cardTitle.appendChild(cardTitleText);
        
        var cardBody = document.createElement('p');
        var cardBodyText = document.createTextNode('This is card content');
        cardBody.appendChild(cardBodyText);
        
        card.appendChild(cardTitle);
        card.appendChild(cardBody);
        grid.appendChild(card);
    }
}

console.log("Created 3x3 card grid");

// ========== 示例 10: 导航菜单 ==========

console.log("\n=== 示例 10: 导航菜单 ===");

var nav = document.createElement('nav');
nav.id = 'mainNav';
nav.className = 'navigation';

var navList = document.createElement('ul');
navList.className = 'nav-list';

var menuItems = ['Home', 'About', 'Services', 'Contact'];

for (var i = 0; i < menuItems.length; i++) {
    var navItem = document.createElement('li');
    navItem.className = 'nav-item';
    
    var navLink = document.createElement('a');
    navLink.setAttribute('href', '#' + menuItems[i].toLowerCase());
    navLink.className = 'nav-link';
    
    var linkText = document.createTextNode(menuItems[i]);
    navLink.appendChild(linkText);
    
    navItem.appendChild(navLink);
    navList.appendChild(navItem);
}

nav.appendChild(navList);

console.log("Navigation menu created with", menuItems.length, "items");

// ========== 示例 11: 查询和遍历 ==========

console.log("\n=== 示例 11: 查询和遍历 ===");

// 查询所有卡片
var allCards = grid.querySelectorAll('.card');
console.log("Found", allCards.length, "cards in grid");

// 查询特定卡片
var specificCard = grid.querySelector('#card-1-1');
if (specificCard) {
    console.log("Found specific card:", specificCard.id);
}

// 查询所有导航链接
var navLinks = nav.querySelectorAll('.nav-link');
console.log("Found", navLinks.length, "navigation links");

// ========== 示例 12: 克隆节点 ==========

console.log("\n=== 示例 12: 克隆节点 ===");

// 克隆按钮（浅克隆）
var buttonClone = button.cloneNode(false);
buttonClone.id = 'clonedButton';
console.log("Cloned button (shallow):", buttonClone.id);

// 克隆容器（深克隆）
var containerClone = container.cloneNode(true);
containerClone.id = 'clonedContainer';
console.log("Cloned container (deep):", containerClone.id);

// ========== 总结 ==========

console.log("\n=== 示例完成 ===");
console.log("演示了以下功能:");
console.log("1. 创建元素和文本节点");
console.log("2. 设置和获取属性");
console.log("3. 构建 DOM 树");
console.log("4. 查询元素 (getElementById, querySelector, querySelectorAll)");
console.log("5. 事件处理和事件冒泡");
console.log("6. 动态修改 DOM");
console.log("7. 创建表单");
console.log("8. 创建网格布局");
console.log("9. 创建导航菜单");
console.log("10. 克隆节点");

